### PR TITLE
Move Chat to Model

### PR DESCRIPTION
**Pull request by TheHollidayInn**

### CODE REVIEW

**Comment by librarianmage** [comment]
If you don't mind me asking, what will this affect?

**Comment by TheHollidayInn** [comment]
This will affect all chat in groups. But this is backwards compatible as I query the chat and apply it to the group model. 

I still have to fix the other routes though before this is done.

**Comment by librarianmage** [comment]
So now chat messages will be forced to conform to this model?

**Comment by TheHollidayInn** [comment]
Yea, I’ll keep what we currently implement and we can add fields. Are you concerned?

**Comment by librarianmage** [comment]
No, just wondering what it will do, since when I did the documentation
about chat, the only info I could go off of was the commented part in the
group model.

File: website/server/models/chat.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+const schema = new mongoose.Schema({
+  timestamp: Date,
+  user: String,
+  text: String,
+  contributor: {type: mongoose.Schema.Types.Mixed},
+  uuid: String,
+  id: String,
+  groupId: {type: String, ref: 'Group'},
+});
+
```
Comment (paglias) [inline website/server/models/chat.js]: We should add the BasePlugin here before merging, like the other models

Comment (TheHollidayInn) [reply website/server/models/chat.js]: Will do!

File: website/server/middlewares/errorHandler.js [hunk 2]
```diff
@@ -16,7 +16,7 @@ module.exports = function errorHandler (err, req, res, next) { // eslint-disable
   // Otherwise try to identify the type of error (mongoose validation, mongodb unique, ...)
   // If we can't identify it, respond with a generic 500 error
   let responseErr = err instanceof CustomError ? err : null;
-
+  console.log(err)
```
Comment (crookedneighbor) [inline website/server/middlewares/errorHandler.js]: I assume this is unintentional.

Comment (TheHollidayInn) [reply website/server/middlewares/errorHandler.js]: Yep. Lint caught it too. I'll remove it during my next updates.

**Comment by TheHollidayInn** [comment]
Fixes #9031

**Comment by negue** [comment]
(I thought first I'd have to run a migration, but it worked out of the box 🎉 )

Will the old messages stay in the group objects or will those be converted to the new model?

**Comment by TheHollidayInn** [comment]
Sweet! The plan is to have the old group messages eventually be removed as new messages come in. But we might just need to migrate them.

**Comment by lemoness** [comment]
@TheHollidayInn What is the status of this?

File: website/server/models/chat.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  flagCount: {type: Number, default: 0},
+  likes: {type: mongoose.Schema.Types.Mixed},
+  userStyles: {type: mongoose.Schema.Types.Mixed},
+}, {
+  minimize: false, // Allow for empty flags to be saved
+});
+
+schema.plugin(baseModel, {
+  noSet: ['_id'],
+  timestamps: true,
```
Comment (paglias) [inline website/server/models/chat.js]: do we care about timestamps here? The message date is already saved in timestamp and I don't think we allow edits

File: website/server/libs/chat/group-chat.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { model as Chat } from '../../models/chat';
+
+// @TODO: Don't use this method when the group can be saved.
+export async function getGroupChat (group) {
+  const groupChat = await Chat.find({groupId: group._id}).limit(200).sort('-timestamp').exec();
```
Comment (paglias) [inline website/server/libs/chat/group-chat.js]: is the limit 400 for some groups?

File: website/server/models/chat.js (hunk 3 above)
Comment (TheHollidayInn) [reply website/server/models/chat.js]: Ah, true. I can remove this.

File: website/server/libs/chat/group-chat.js (hunk 4 above)
Comment (TheHollidayInn) [reply website/server/libs/chat/group-chat.js]: Not that I know of. We talked about doing it for Group Plans, but after we merged this PR.

Comment (Alys) [reply website/server/libs/chat/group-chat.js]: I think it's already in place.
website/server/models/group.js:537:  const MAX_SUBBED_GROUP_CHAT_COUNT = 400;

Comment (TheHollidayInn) [reply website/server/libs/chat/group-chat.js]: Ah, okay. I can change that.

**Comment by paglias** [comment]
When this gets merged, we should post a message in #10019

**Comment by Alys** [comment]
When a quest starts on production, the message in chat is like this:
```
        {
            "id": "c2819113-708e-4c85-b74e-64f19118c15b",
            "text": "`Your quest, Attack of the Mundane, Part 1: Dish Disaster!, has started.`",
            "timestamp": 1498284774350,
            "likes": {},
            "flags": {},
            "flagCount": 0,
            "uuid": "system",
            "_meta": {
                "participatingMembers": "d904bd62-da08-416b-a816-ba797c9ee265"
            }
        },
```

In beta it's like this:
```
{
	"_id" : "a1e756e8-3bdf-4eef-8d36-80f8847dd7b6",
	"flagCount" : 0,
	"flags" : {
		
	},
	"id" : "45a931d1-05ed-40ab-9e6d-2f2a7e25bf42",
	"text" : "`Your quest, The Basi-List, has started.`",
	"timestamp" : ISODate("2018-04-19T11:00:30.485Z"),
	"likes" : {
		
	},
	"uuid" : "system",
	"groupId" : "ae9bf3a2-8050-4253-85ba-14eab6b19e72",
	"__v" : 0
}
```

I.e., it's missing the `_meta` object.

Also, there's tests for that object in `develop`, but they haven't told us about this error, so those tests must be broken in beta. That's worrying so maybe they could be fixed before the `_meta` object is added back into the chat messages so that we can be sure those tests are working?

**Comment by TheHollidayInn** [comment]
I have that fixed on the PR. I'll update Beta in a sec

**Comment by Alys** [comment]
Old guild messages are not being purged when there's more than 200 of them. If you'd like to test, you can use guild 4e5f91ea-692e-44d4-bd8d-82201ddc1313 which has slightly over 200 messages now, most of them numbered - feel free to post whatever you want there. The same should apply to parties and the Tavern. For any guild or party that has a group plan, the number of messages kept must be 400.

**Comment by Alys** [comment]
For the migration script that will convert messages from the old storage system to the new, can I get a copy of it to test on a local install before it's run over the beta database? It will be easier to confirm that it's working correctly locally where I have only a small number of guilds and parties. Once that's tested, it could be run over the beta database and then I can do some final tests (e.g., to confirm that the console errors I've been seeing are just from messages existing in the old format).

**Comment by Alys** [comment]
Mostly for my own reference so I don't forget:

We must remember to make sure the Blacksmiths know about that migration script too. Many of them are likely to have local installs set up with a lot of test data (I know I do) and it would be convenient for them to be able to convert their database rather than having to start from scratch.

**Comment by TheHollidayInn** [comment]
- Just pushed the limit fix. The vars weren't being exported - I missed that :P  
- I pushed the migration. However, I have already run it on Beta Thursday or Friday.
- Console message like this `TypeError: Cannot read property 'classSelected' of undefined` and some other user errors are because I moved old prod chat over but didn't move every user over. Beta has a migration to grab userIds and move a lot of that connected user data, but not all. So that leads to chat messages with non existing users.

File: migrations/groups/migrate-chat.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
+// @migrationName = 'MigrateGroupChat';
+// @authorName = 'TheHollidayInn'; // in case script author needs to know when their ...
+// @authorUuid = ''; // ... own data is done
+
+
+/*
+ * This migrations will iterate through all groups with a group plan a subscription and resync the free
+ * subscription to all members
```
Comment (Alys) [inline migrations/groups/migrate-chat.js]: Can this comment be updated?

**Comment by Alys** [comment]
The console log shows a lot of `TypeError: e.flags is undefined` errors in the Tavern and parties. To see them, visit one of those pages and then reload it. They don't happen in production. There might be other errors as well but they're hard to find so even if the e.flags errors are benign, it would be good if they could be prevented.

In the 5481ccf3-5d2d-48a9-a871-70a7380cee5a guild, I saw two examples of "t.response is undefined" in the console after one reload but I can't replicate it so it might have been a temporary problem and if you want to ignore it, that's fine with me. :)
```
Unhandled promise rejection 
TypeError
columnNumber: 1
fileName: "https://habitrpg-beta.herokuapp.com/static/js/app.d2e6be8e3bb70e4bebb7.js"
lineNumber: 1
message: "t.response is undefined"
stack: "created/<@https://habitrpg-beta.herokuapp.com/static/js/app.d2e6be8e3bb70e4bebb7.js:1:1\nT/</<@https://habitrpg-beta.herokuapp.com/static/js/vendor.b91cfcde844f24cdc5d8.js:6:211261\nT/<@https://habitrpg-beta.herokuapp.com/static/js/vendor.b91cfcde844f24cdc5d8.js:6:211131\nl@https://habitrpg-beta.herokuapp.com/static/js/vendor.b91cfcde844f24cdc5d8.js:6:74862\n"
```

I'm getting poor performance in my browser when viewing guilds/party/tavern on beta, with my browser really struggling when I reload the site, but that might just be because it's the beta server. Let me know if you want any more details.

Can this line https://github.com/HabitRPG/habitica/blob/develop/migrations/migration-runner.js#L20 be changed as part of this PR? All of the contributors will need to run that script over their own databases and it will be much easier to explain to them how if we don't have to tell them to edit that script and then revert the edit when they're finished. :)
FYI I'll document on the wiki how that migration-runner script works and how normally you do need to edit it to run any migration in general, but for this one specific case it would be good for them to not have to edit.

**Comment by Alys** [comment]
I still can't get old messages to automatically delete. Guild 4e5f91ea-692e-44d4-bd8d-82201ddc1313 has more than 200 now and when I post more, the old ones don't disappear.

File: website/server/controllers/api-v3/chat.js [hunk 6]
```diff
@@ -164,35 +166,35 @@ api.postChat = {
...
+      user.party.lastMessageSeen = newChatMessage.id;
       toSave.push(user.save());
     }
 
-    let [savedGroup] = await Promise.all(toSave);
+    await Promise.all(toSave);
 
-    // realtime chat is only enabled for private groups (for now only for parties)
-    if (savedGroup.privacy === 'private' && savedGroup.type === 'party') {
+    // real-time chat is only enabled for private groups (for now only for parties)
```
Comment (Alys) [inline website/server/controllers/api-v3/chat.js]: Not exactly related to this PR, but should this comment be changed to something like "real-time chat is not currently running" or "TODO make real-time chat work" or something along those lines?

**Pull request merged**